### PR TITLE
Fix/GC big objects parts

### LINKED ISSUE #2858: GC has not collected an expired object
## Expected Behavior
Expired objects are deleted. Even if the node was down for some time, it should delete them afterwards.

## Current Behavior
```
May 30 14:54:27 metis2 neofs-node[656]: 2024-05-30T14:54:27.505Z        error        replicator/process.go:76        could not replicate object        {"component": "Object Replicator", "node": "03aeff8a19f0202090afb0916b1c00b432321be7e8623a06c9b9b5db8ee5c053a4", "object": "HXSaMJXk2g8C14ht8HSi7BBaiYZ1HeWh2xnWPGQCg4H6/6Ha2WhpvBsAeT23jsa9Ak8DVrqVB97xLXTSeLrjYnoGD", "error": "copy object using NeoFS API client of the remote node: status: code = 1024 message = failed to verify and store object locally: validate object format: object did not pass expiration check: object has expired"}
```
It tries to replicate the same set of objects again and again which means GC has failed to do its job.

## Possible Solution
Fix GC.

## Steps to Reproduce (for bugs)
Shut an existing node down for some time and expand it with one shard (3->4).

## Your Environment
* Version used: 0.42.0
* Server setup and configuration: mainnet
* Operating System and version (`uname -a`): Debian stable

### PR DESCRIPTION
**Pull request by carpawell**
Closes #2858.

### CODE REVIEW

**Comment by carpawell** [comment]
Found corner case in shard distribution, will add tests and fix.

**Comment by carpawell** [comment]
OK, a fresh view and an expended test showed that it did not work. Let it be simple and error-proof.

File: pkg/core/object/object.go [hunk 1]
```diff
@@ -26,3 +27,20 @@ func AddressOf(obj *object.Object) oid.Address {
 
 	return addr
 }
+
+// ErrNoExpiration means no expiration was set.
+var ErrNoExpiration = errors.New("missing expiration epoch attribute")
+
+// Expiration searches for expiration attribute in the object. Returns
+// ErrNoExpiration if not found.
+func Expiration(obj *object.Object) (uint64, error) {
```
Comment (cthulhu-rider) [inline pkg/core/object/object.go:36]:
```suggestion
func Expiration(obj object.Object) (uint64, error) {
```

File: pkg/local_object_storage/engine/engine_test.go [hunk 2]
```diff
@@ -28,10 +29,12 @@ import (
...
 )
 
-type epochState struct{}
+type epochState struct {
+	e uint64
+}
 
-func (s epochState) CurrentEpoch() uint64 {
-	return 0
+func (s *epochState) CurrentEpoch() uint64 {
```
Comment (cthulhu-rider) [inline pkg/local_object_storage/engine/engine_test.go:36]: why switched to pointer?

File: pkg/services/replicator/process.go [hunk 3]
```diff
@@ -45,6 +49,11 @@ func (p *Replicator) HandleTask(ctx context.Context, task Task, res TaskResult)
 			stream = client.DemuxReplicatedObject(stream)
 		}
 	} else {
+		if err = p.checkExpiration(task.obj); err != nil {
```
Comment (cthulhu-rider) [inline pkg/services/replicator/process.go:52]: minor: this is not Replicator's concern imo, i'd make such exclusions on caller side

File: pkg/services/replicator/process.go [hunk 4]
```diff
@@ -45,6 +49,11 @@ func (p *Replicator) HandleTask(ctx context.Context, task Task, res TaskResult)
...
+			p.log.Debug("expiration check", zap.Error(err))
+			return
```
Comment (cthulhu-rider) [inline pkg/services/replicator/process.go:54]: iiuc `checkExpiration` eternal objects w/o parent are never replicated. I doubt this is a correct behavior

File: pkg/services/replicator/process.go [hunk 5]
```diff
@@ -85,3 +94,24 @@ func (p *Replicator) HandleTask(ctx context.Context, task Task, res TaskResult)
 		}
 	}
 }
+
+func (p *Replicator) checkExpiration(o *objectSDK.Object) error {
```
Comment (cthulhu-rider) [inline pkg/services/replicator/process.go:98]:
```suggestion
func (p *Replicator) checkExpiration(o objectSDK.Object) error {
```

File: pkg/services/replicator/process.go [hunk 6]
```diff
@@ -45,6 +49,11 @@ func (p *Replicator) HandleTask(ctx context.Context, task Task, res TaskResult)
 			stream = client.DemuxReplicatedObject(stream)
 		}
 	} else {
+		if err = p.checkExpiration(task.obj); err != nil {
+			p.log.Debug("expiration check", zap.Error(err))
```
Comment (cthulhu-rider) [inline pkg/services/replicator/process.go:53]: i'd `Info` (at least) cases with broken exp attributes cuz we don't expect them in general

Comment (carpawell) [reply pkg/services/replicator/process.go]: ok, added details and made it warn

File: pkg/services/replicator/process.go (hunk 4 above)
Comment (carpawell) [reply pkg/services/replicator/process.go:54]: fixed

File: pkg/services/replicator/process.go (hunk 3 above)
Comment (carpawell) [reply pkg/services/replicator/process.go]: objects are taken in a batch and handled via a pool of _some size_ in the Policer, so Replicator to me is the closest thing to the PUT/REPLICATE operation and is the best thing to abort operation if an object is expired b/w engine's SELECT and network replication. but i understand you, can move it to Policer @roman-khimov

File: pkg/local_object_storage/engine/engine_test.go (hunk 2 above)
Comment (carpawell) [reply pkg/local_object_storage/engine/engine_test.go]: rework artifact, rolled back

File: pkg/services/replicator/process.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/services/replicator/process.go]: Caller doesn't have object header to judge on this IIUC. And replicator has to get it anyway.

Comment (roman-khimov) [reply pkg/services/replicator/process.go]: Ah, it can have it in this case. Then policer would be a better place to keep this code doing exactly one thing.

Comment (carpawell) [reply pkg/services/replicator/process.go]: @roman-khimov, @cthulhu-rider, i rechecked, a full obj can be here only in evacuation case, in a normal replication case it is now binary replication, so it is impossible to get expiration (or make additional unmarshaling after we make replication without unmarshaling 😋). so only evacuation can be improved (good, can be done but is almost useless to me), i would just drop this commit then (evacuation can be improved, but this is a control function, so even no logs can be done if smth is skipped)

Comment (carpawell) [reply pkg/services/replicator/process.go]: dropped commit

File: pkg/core/object/object.go [hunk 7]
```diff
@@ -26,3 +27,20 @@ func AddressOf(obj *object.Object) oid.Address {
 
 	return addr
 }
+
+// ErrNoExpiration means no expiration was set.
+var ErrNoExpiration = errors.New("missing expiration epoch attribute")
+
+// Expiration searches for expiration attribute in the object. Returns
+// ErrNoExpiration if not found.
+func Expiration(obj object.Object) (uint64, error) {
```
Comment (roman-khimov) [inline pkg/core/object/object.go:36]: Do we need this if we're not touching replicator?

**Review by cthulhu-rider — APPROVED** [review]

File: pkg/core/object/object.go (hunk 7 above)
Comment (carpawell) [reply pkg/core/object/object.go:36]: this can be useful in general as i mentioned in the commit, i just had not searched if more places can reused it
